controller/scheduler-v2/scheduler.go

```diff
@@ -69,62 +80,91 @@ func (s *Scheduler) Run() error {
 	log.Info("starting scheduler loop")
 	defer log.Info("exiting scheduler loop")
 
-	jobTicker := time.Tick(30 * time.Second)
-	formationTicker := time.Tick(time.Minute)
-	formationTimeout := time.After(time.Second)
-	go func() {
-		for {
-			select {
-			case <-jobTicker:
-				triggerChan(s.jobSync)
-			case <-formationTicker:
-				triggerChan(s.formationSync)
-			case <-formationTimeout:
-				triggerChan(s.formationSync)
-			}
-		}
-	}()
-	triggerChan(s.jobSync)
+	tickChannel(s.syncJobs, 30*time.Second)
+	tickChannel(s.syncFormations, time.Minute)
+	tickChannel(s.syncHosts, 5*time.Minute)
+
+	stream, err := s.StreamHosts(s.hostChange)
+	if err != nil {
+		return fmt.Errorf("Unable to stream hosts. Error: %v", err)
+	}
+	defer stream.Close()
+
+	stream, err = s.StreamFormations(nil, s.formationChange)
+	if err != nil {
+		return fmt.Errorf("Unable to stream formations. Error: %v", err)
+	}
+	defer stream.Close()
+
+	go s.RunPutJobs()
 
 	for {
+		// First handle events reconciling our state with the cluster
 		select {
 		case <-s.stop:
 			return nil
-		case req := <-s.jobRequests:
-			s.jobRequestHandler(req)
-		case <-s.rectifyJobs:
-			s.RectifyJobs()
+		case h := <-s.hostChange:
+			s.followHost(h)
+			continue
+		case he := <-s.hostEvents:
+			s.handleHostEvent(he)
+			continue
 		case ef := <-s.formationChange:
-			s.formationChangeHandler(ef)
-		case <-s.formationSync:
+			s.FormationChange(ef)
+			continue
+		case <-time.After(50 * time.Millisecond):
```

Comment:
Ok I get that, but the issue with where this is at the moment is if a formation change event comes in and 10 jobs need to be started, we are going to sleep 50ms between starting each one (and maybe even longer as we receive host events for those jobs), so if this is needed it should be at the end of the loop when we know there is no work to do.

Also, given I asked the question, I would put a comment next to it (arbitrary sleeps always look suspicious).